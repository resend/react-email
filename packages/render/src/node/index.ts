import type { Options } from '../shared/options';
import { render } from './render';

/**
 * @deprecated use {@link render}
 */
export const renderAsync = (element: React.ReactElement, options?: Options) => {
  return render(element, options);
};

export * from '../shared/options';
export * from '../shared/utils/pretty';
export * from '../shared/utils/to-plain-text';
export * from '../shared/utils/to-resend-template';
export * from './render';
