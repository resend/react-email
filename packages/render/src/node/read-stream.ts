import { Writable } from 'node:stream';
import type {
  PipeableStream,
  ReactDOMServerReadableStream,
} from 'react-dom/server.browser';

const decoder = new TextDecoder('utf-8');

export const readStream = async (
  stream: PipeableStream | ReactDOMServerReadableStream,
) => {
  let result = '';

  if ('pipeTo' in stream) {
    // means it's a readable stream
    const writableStream = new WritableStream({
      write(chunk: BufferSource) {
        result += decoder.decode(chunk);
      },
    });
    await stream.pipeTo(writableStream);
  } else {
    const writable = new Writable({
      write(chunk: BufferSource, _encoding, callback) {
        result += decoder.decode(chunk);

        callback();
      },
    });
    stream.pipe(writable);

    await new Promise<void>((resolve, reject) => {
      writable.on('error', reject);
      writable.on('close', () => {
        resolve();
      });
    });
  }

  return result;
};
