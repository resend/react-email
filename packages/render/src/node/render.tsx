import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';
import { readStream } from './read-stream';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
