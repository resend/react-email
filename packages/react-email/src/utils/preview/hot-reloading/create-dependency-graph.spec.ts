import { existsSync, promises as fs } from 'node:fs';
import path from 'node:path';
import {
  createDependencyGraph,
  type DependencyGraph,
} from './create-dependency-graph.js';

const pathToFileForTestingDependencyGraph = path.join(
  __dirname,
  '.for-dependency-graph.ts',
);

vi.mock('@babel/traverse', async () => {
  const traverse = await vi.importActual('@babel/traverse');
  return { default: traverse };
});

test('createDependencyGraph()', async () => {
  if (existsSync(pathToFileForTestingDependencyGraph)) {
    await fs.rm(pathToFileForTestingDependencyGraph);
  }

  const [dependencyGraph, updateDependencyGraph] =
    await createDependencyGraph(__dirname);

  const toAbsolute = (relativePath: string) => {
    return path.resolve(__dirname, relativePath);
  };

  const convertPathsToAbsolute = (graph: DependencyGraph) => {
    return Object.fromEntries(
      Object.entries(graph).map(([relativePath, module]) => {
        return [
          toAbsolute(relativePath),
          {
            path: toAbsolute(relativePath),
            dependencyPaths: module.dependencyPaths.map((p) => toAbsolute(p)),
            dependentPaths: module.dependentPaths.map((p) => toAbsolute(p)),
            moduleDependencies: module.moduleDependencies,
          },
        ];
      }),
    );
  };

  const initialDependencyGraph = convertPathsToAbsolute({
    'create-dependency-graph.ts': {
      path: 'create-dependency-graph.ts',
      dependencyPaths: ['get-imported-modules.ts', 'resolve-path-aliases.ts'],
      dependentPaths: [
        'create-dependency-graph.spec.ts',
        'setup-hot-reloading.ts',
      ],
      moduleDependencies: ['node:fs', 'node:path', 'chokidar/handler.js'],
    },
    'create-dependency-graph.spec.ts': {
      path: 'create-dependency-graph.spec.ts',
      dependencyPaths: ['create-dependency-graph.ts'],
      dependentPaths: [],
      moduleDependencies: ['node:fs', 'node:path'],
    },
    './test/some-file.ts': {
      dependencyPaths: [],
      dependentPaths: [],
      moduleDependencies: [],
      path: '/home/gabriel/Projects/Resend/react-email/packages/react-email/src/cli/utils/preview/hot-reloading/test/some-file.ts',
    },
    'resolve-path-aliases.ts': {
      path: 'resolve-path-aliases.ts',
      dependentPaths: [
        'create-dependency-graph.ts',
        'resolve-path-aliases.spec.ts',
      ],
      dependencyPaths: [],
      moduleDependencies: ['node:path', 'tsconfig-paths'],
    },
    'resolve-path-aliases.spec.ts': {
      path: 'resolve-path-aliases.spec.ts',
      dependencyPaths: ['resolve-path-aliases.ts'],
      dependentPaths: [],
      moduleDependencies: ['node:path'],
    },
    'get-imported-modules.ts': {
      path: 'get-imported-modules',
      dependentPaths: [
        'create-dependency-graph.ts',
        'get-imported-modules.spec.ts',
      ],
      dependencyPaths: [],
      moduleDependencies: ['@babel/parser', '@babel/traverse'],
    },
    'get-imported-modules.spec.ts': {
      path: 'get-imported-modules.spec.ts',
      dependencyPaths: ['get-imported-modules.ts'],
      dependentPaths: [],
      moduleDependencies: ['node:fs'],
    },
    'setup-hot-reloading.ts': {
      path: 'setup-hot-reloading.ts',
      dependencyPaths: [
        '../../types/hot-reload-change.ts',
        'create-dependency-graph.ts',
      ],
      dependentPaths: [],
      moduleDependencies: [
        'node:http',
        'node:path',
        'chokidar',
        'debounce',
        'socket.io',
      ],
    },
    '../../types/hot-reload-event.ts': {
      dependencyPaths: [],
      dependentPaths: ['../../types/hot-reload-change.ts'],
      moduleDependencies: ['chokidar/handler.js'],
      path: '../../types/hot-reload-event.ts',
    },
    '../../types/hot-reload-change.ts': {
      path: '../../types/hot-reload-change.ts',
      dependencyPaths: ['../../types/hot-reload-event.ts'],
      dependentPaths: ['setup-hot-reloading.ts'],
      moduleDependencies: [],
    },
  } satisfies DependencyGraph);

  expect(
    dependencyGraph,
    'the initial value for the dependency graph should work with the directory of this testing file',
  ).toEqual(initialDependencyGraph);

  await fs.writeFile(
    pathToFileForTestingDependencyGraph,
    `
import {} from './setup-hot-reloading';
import {} from './get-imported-modules';
import {} from './create-dependency-graph.ts';
`,
    'utf8',
  );
  await updateDependencyGraph('add', pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[pathToFileForTestingDependencyGraph],
    'added file to have proper dependency paths',
  ).toEqual({
    path: pathToFileForTestingDependencyGraph,
    dependentPaths: [],
    dependencyPaths: [
      toAbsolute('setup-hot-reloading.ts'),
      toAbsolute('get-imported-modules.ts'),
      toAbsolute('create-dependency-graph.ts'),
    ],
    moduleDependencies: [],
  } satisfies DependencyGraph[number]);
  expect(
    dependencyGraph[toAbsolute('setup-hot-reloading.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('get-imported-modules.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('create-dependency-graph.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);

  await fs.writeFile(
    pathToFileForTestingDependencyGraph,
    `
import {} from './setup-hot-reloading';
import {} from './create-dependency-graph.ts';
`,
    'utf8',
  );
  await updateDependencyGraph('change', pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[pathToFileForTestingDependencyGraph],
    'changed file to have updated dependencyPaths',
  ).toEqual({
    path: pathToFileForTestingDependencyGraph,
    dependentPaths: [],
    dependencyPaths: [
      toAbsolute('setup-hot-reloading.ts'),
      toAbsolute('create-dependency-graph.ts'),
    ],
    moduleDependencies: [],
  } satisfies DependencyGraph[number]);
  expect(
    dependencyGraph[toAbsolute('setup-hot-reloading.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('get-imported-modules.ts')]?.dependentPaths,
    'when removing dependency on a file, the dependency should have its dependents updated to not have the testing file again',
  ).not.toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('create-dependency-graph.ts')]?.dependentPaths,
  ).toContain(pathToFileForTestingDependencyGraph);

  await fs.rm(pathToFileForTestingDependencyGraph);
  await updateDependencyGraph('unlink', pathToFileForTestingDependencyGraph);
  expect(dependencyGraph[pathToFileForTestingDependencyGraph]).toBeUndefined();
  expect(
    dependencyGraph[toAbsolute('setup-hot-reloading.ts')]?.dependentPaths,
    "should remove itself from dependents once it's unlinked",
  ).not.toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('get-imported-modules.ts')]?.dependentPaths,
    "should remove itself from dependents once it's unlinked",
  ).not.toContain(pathToFileForTestingDependencyGraph);
  expect(
    dependencyGraph[toAbsolute('create-dependency-graph.ts')]?.dependentPaths,
    "should remove itself from dependents once it's unlinked",
  ).not.toContain(pathToFileForTestingDependencyGraph);
});
