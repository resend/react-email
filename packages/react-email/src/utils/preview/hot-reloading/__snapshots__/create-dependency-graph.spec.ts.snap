// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDependencyGraph() > should have the right initial value for the dependency graph 1`] = `
{
  "../outer.ts": {
    "dependencyPaths": [
      "outer-dependency.ts",
    ],
    "dependentPaths": [
      "general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "../outer.ts",
  },
  "data-to-import.json": {
    "dependencyPaths": [],
    "dependentPaths": [
      "general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "data-to-import.json",
  },
  "file-a.ts": {
    "dependencyPaths": [
      "file-b.ts",
    ],
    "dependentPaths": [
      "file-b.ts",
      "general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "file-a.ts",
  },
  "file-b.ts": {
    "dependencyPaths": [
      "file-a.ts",
    ],
    "dependentPaths": [
      "file-a.ts",
      "general-importing-file.ts",
    ],
    "moduleDependencies": [],
    "path": "file-b.ts",
  },
  "general-importing-file.ts": {
    "dependencyPaths": [
      "../outer.ts",
      "data-to-import.json",
      "file-a.ts",
      "file-b.ts",
    ],
    "dependentPaths": [],
    "moduleDependencies": [
      "node:os",
      "node:path",
    ],
    "path": "general-importing-file.ts",
  },
  "outer-dependency.ts": {
    "dependencyPaths": [],
    "dependentPaths": [
      "../outer.ts",
    ],
    "moduleDependencies": [],
    "path": "outer-dependency.ts",
  },
}
`;
